src/js/lib/routing.ts

```diff
@@ -0,0 +1,58 @@
+import { toggleCardDetails } from './utils';
+
+/**
+ * Pushes the card ID to the history stack. Prevents the page from jumping
+ * as it would happen when setting the location.hash.
+ * @param {string} id - The ID of the card
+ */
+export function pushCardState(id: string): void {
+  history.pushState({ id }, null, `#${id}`);
+}
+
+/**
+ * Open the initially selected card and listen for popstate events.
+ */
+export function registerRouting(): void {
+  openCardFromHash();
+  window.addEventListener('popstate', (event: PopStateEvent) => restoreState(event));
+}
+
+/**
+ * Opens the card from the initial ID in the hash (if present).
+ */
+function openCardFromHash(): void {
+  // Open card from hash
+  const cardId = location.hash.substr(1); // strip #
+  const target = document.getElementById(cardId);
+  if (target) {
+    toggleCardDetails(target);
+  }
+}
+
+/**
+ * State updated due to change in browser history. Opens the card if necessary
+ * and aligns it to the top of the page.
+ * @param {PopStateEvent} event
+ */
+function restoreState(event: PopStateEvent): void {
```

Comment:
There's a bug with this if filters are applied; if a filter is applied and then the hash, the resulting URL will still open the correct card but it won't get scrolled into view. If, on the other hand, the hash is first and then the query parameters for the filter, it won't do either.